synapse/module_api/__init__.py

```diff
@@ -73,6 +87,62 @@ def public_room_list_manager(self):
         """
         return self._public_room_list_manager
 
+    def register_hook(self, fn_name: str, callback: Callable):
+        """Registers the given callback as a hook with the given name.
+
+        This function must be called in the __init__ method of the modules using it.
+
+        If multiple modules register the same hook, their callbacks will be run in the
+        same order as the modules appear in the configuration file (highest first).
+
+        Args:
+            fn_name: The hook name.
+            callback: The function to call for this hook.
+        """
+        if fn_name not in self._callbacks.keys():
+            self._callbacks[fn_name] = [callback]
+            return
+
+        self._callbacks[fn_name].append(callback)
+
+    def get_registered_hook(self, fn_name: str) -> List[Callable]:
```

Comment:
Currently these look like they could be used by modules, even though the intent is very much that they're called from within Synapse, I'm not sure how to make this clear enough (maybe changing them so they're private methods? feels a bit hacky since we'll need to call them from outside the `ModuleApi`).

Reply:
well the obvious answer is to move `_callbacks` and `get_registered_hook` somewhere else, and have `register_hook` proxy to that place.

However, I'm dubious about having a single function for all hooks and callbacks. Better to have individual functions for each type of hook - for example you might have `ModuleApi.register_spamcheck_callback`, which proxies to a method on `hs.get_spam_checker()`, and the `SpamChecker` is responsible for maintaining its own list of callbacks.

A big advantage of doing so is that you can make the definition more typesafe - ie, declare it to take a `Callable[[Event], bool]` rather than just a `Callable`.

Reply:
> A big advantage of doing so is that you can make the definition more typesafe - ie, declare it to take a Callable[[Event], bool] rather than just a Callable.

Not necessarily because all e.g. spamchecker callbacks don't have the same signature, unless you're proposing we create a `register_` method for each hook which doesn't sound like the right solution.

Overall it does sound like a better solution though, thanks.

Reply:
Implemented in https://github.com/matrix-org/synapse/pull/10062/commits/f5098c9113e5d20ffdc8ef0abe3a09f33795773c